### PR TITLE
Update secp256k keys with recovery id

### PR DESCRIPTION
**Pull request by batconjurer**
This ports the changes from #617 to our branch and updates the key files to be the same as main.

### CODE REVIEW

File: shared/src/types/key/mod.rs [hunk 1]
```diff
@@ -356,6 +357,26 @@ impl<PK: PublicKey> From<&PK> for PublicKeyHash {
...
+            pkh.0
+        }
+        common::PublicKey::Secp256k1(pk) => {
+            let pkh = PublicKeyHash::from(pk);
+            pkh.0
+        }
+    }
+}
+
+/// Convert Tendermint validator's raw hash bytes to Anoma raw hash string
```
Comment (sug0) [inline shared/src/types/key/mod.rs:375]:
```suggestion
/// Convert Tendermint validator's raw hash bytes to Namada raw hash string
```

File: shared/src/types/key/secp256k1.rs [hunk 2]
```diff
@@ -316,7 +323,7 @@ impl<'de> Deserialize<'de> for Signature {
         struct ByteArrayVisitor;
 
         impl<'de> Visitor<'de> for ByteArrayVisitor {
-            type Value = [u8; libsecp256k1::util::SIGNATURE_SIZE];
+            type Value = [u8; libsecp256k1::util::SIGNATURE_SIZE + 1];
```
Comment (sug0) [inline shared/src/types/key/secp256k1.rs:326]: this has probably already gone into main, but we could add a const here with a docstr explaining why this `+ 1` is needed

**Review by sug0 — CHANGES_REQUESTED** [review]

**Comment by batconjurer** [comment]
pls update wasm

File: shared/src/types/key/mod.rs [hunk 3]
```diff
@@ -356,6 +357,26 @@ impl<PK: PublicKey> From<&PK> for PublicKeyHash {
...
+            let pkh = PublicKeyHash::from(pk);
+            pkh.0
+        }
+    }
+}
+
+/// Convert Tendermint validator's raw hash bytes to Namada raw hash string
+pub fn tm_raw_hash_to_string(raw_hash: impl AsRef<[u8]>) -> String {
+    HEXUPPER.encode(raw_hash.as_ref())
+}
```
Comment (james-chf) [inline shared/src/types/key/mod.rs:378]: I'm not sure about these helpers here, could they live with the Tendermint related code that will use them? (they aren't used yet in this PR). These might belong in a separate PR

File: shared/src/types/key/mod.rs [hunk 4]
```diff
@@ -356,6 +357,26 @@ impl<PK: PublicKey> From<&PK> for PublicKeyHash {
...
+            pkh.0
+        }
+        common::PublicKey::Secp256k1(pk) => {
+            let pkh = PublicKeyHash::from(pk);
+            pkh.0
+        }
+    }
+}
+
+/// Convert Tendermint validator's raw hash bytes to Namada raw hash string
```
Comment (james-chf) [inline shared/src/types/key/mod.rs:375]: `validator's raw hash bytes` - not sure what this is, test/example would be useful here

File: shared/src/types/key/secp256k1.rs [hunk 5]
```diff
@@ -316,7 +323,10 @@ impl<'de> Deserialize<'de> for Signature {
         struct ByteArrayVisitor;
 
         impl<'de> Visitor<'de> for ByteArrayVisitor {
-            type Value = [u8; libsecp256k1::util::SIGNATURE_SIZE];
+            // [`libsecp265k1::util::SIGNATURE_SIZE`] is for a traditional
+            // signature on this curve. For Ethereum, an extra byte is included
+            // that prevents malleability attacks.
+            type Value = [u8; libsecp256k1::util::SIGNATURE_SIZE + 1];
```
Comment (james-chf) [inline shared/src/types/key/secp256k1.rs:329]: The `libsecp256k1::util::SIGNATURE_SIZE + 1` and comment should be pulled out into a const and docstring e.g. `SIG_BYTES_LEN`, then whereever we are using `[u8; 65]`/`libsecp256k1::util::SIGNATURE_SIZE + 1` or similar we should use the const instead e.g. `[u8; SIG_BYTES_LEN]`.

File: shared/src/types/key/secp256k1.rs [hunk 6]
```diff
@@ -304,6 +310,7 @@ impl Serialize for Signature {
         for elem in &arr[..] {
```
Comment (james-chf) [inline shared/src/types/key/secp256k1.rs:310]: On L309
```
        let mut seq = serializer.serialize_tuple(arr.len())?;
```

should this be `arr.len() + 1` now that we have the extra `seq.serialize_element()` call? We should test our Serialize/Deserialize impls match up (we should already have had tests probably, I could have a go at this even as a separate PR)

**Review by james-chf — COMMENTED** [review]
have started taking a look, will review https://github.com/anoma/namada/pull/579 and other PRs before this for context

File: shared/src/types/key/mod.rs (hunk 3 above)
Comment (batconjurer) [reply shared/src/types/key/mod.rs:378]: Those got grabbed when I synced this file with main. There was a bug when looking up tendermint addresses (@sug0  should remember. Something about not being able to slash if someones address wasn't valid utf-8). I can remove these as they aren't directly relevant to this PR, but they will reappear if we ever merge in main.

File: shared/src/types/key/mod.rs (hunk 4 above)
Comment (batconjurer) [reply shared/src/types/key/mod.rs:375]: Same thing, this got pulled in from main. Can remove if necessary

File: shared/src/types/key/secp256k1.rs (hunk 6 above)
Comment (batconjurer) [reply shared/src/types/key/secp256k1.rs:310]: good catch! Will change and cover with tests.

File: shared/src/types/key/mod.rs (hunk 3 above)
Comment (james-chf) [reply shared/src/types/key/mod.rs:378]: Oh right, I thought these were being added new for some further PR, if they are already merged in `main` then I guess it's immaterial. I'm guessing the bug was re. https://github.com/anoma/namada/blob/a669ca573f1d3f61b87cdcb8eba7cb9a6f872a6a/shared/src/types/vote_extensions/ethereum_events.rs#L24-L27

Comment (james-chf) [reply shared/src/types/key/mod.rs:378]: I've opened https://github.com/anoma/namada/issues/635. I don't fully recall the issue either but I think it was that a validator should be able to look at it's own storage to get a validator's address, rather than trust an address provided in vote extensions.

Comment (sug0) [reply shared/src/types/key/mod.rs:378]: we were trying to parse utf8 data from raw hashes, which are of course completely arbitrary, and there was a piece of code that effectively skipped slashing if you failed to parse a hash as utf8 data.

**Comment by sug0** [comment]
ci is throwing a fit. could we fix these issues before merging?

**Comment by batconjurer** [comment]
pls update wasm

**Comment by batconjurer** [comment]
pls update wasm

File: shared/src/types/key/secp256k1.rs [hunk 7]
```diff
@@ -405,12 +435,19 @@ impl PartialOrd for Signature {
...
-impl TryFrom<&[u8; 64]> for Signature {
+impl TryFrom<&[u8; 65]> for Signature {
     type Error = ParseSignatureError;
 
-    fn try_from(sig: &[u8; 64]) -> Result<Self, Self::Error> {
-        libsecp256k1::Signature::parse_standard(sig)
-            .map(Self)
+    fn try_from(sig: &[u8; 65]) -> Result<Self, Self::Error> {
+        let sig_bytes = sig[..64].try_into().unwrap();
+        let recovery_id = RecoveryId::parse(sig[64]).map_err(|err| {
```
Comment (james-chf) [inline shared/src/types/key/secp256k1.rs:443]: We could use `SIGNATURE_LENGTH`/`SIGNATURE_LENGTH - 1` here as well instead of `65`/`64`

**Review by james-chf — APPROVED** [review]
lgtm
